File: 2017/fall/3/binary0.c
// Iterative binary search

#include <cs50.h>
#include <stdio.h>
#include <string.h>

// Names in a phone book
string book[] = {
    "Chen",
    "Kernighan",
    "Leitner",
    "Lewis",
    "Malan",
    "Muller",
    "Seltzer",
    "Shieber",
    "Smith"
};

bool search(string name, string names[], int left, int right);

int main(void)
{
    // Prompt user for name
    string name = get_string("Name: ");

    // Search for name
    if (search(name, book, 0, sizeof(book) / sizeof(string) - 1))
    {
        printf("Calling %s\n", name);
    }
    else
    {
        printf("Quitting\n");
    }
}

// Searches names for name
bool search(string name, string names[], int left, int right)
{
    // No more names to search
    if (left > right)
    {
        return false;
    }

    // Look at middle
    int middle = (left + right) / 2;
    if (strcmp(name, names[middle]) == 0)
    {
        return true;       
    }

    // Search left half
    else if (strcmp(name, names[middle]) < 0)
    {
        return search(name, names, left, middle - 1);
    }

    // Search right half
    else if (strcmp(name, names[middle]) > 0)
    {
        return search(name, names, middle + 1, right);
    }

    return false;
}

File: 6/src6/weeks/3/binary0.py
# Iterative binary search

import sys
from cs50 import get_string

# Names in a phone book
book = [
    "Chen",
    "Kernighan",
    "Leitner",
    "Lewis",
    "Malan",
    "Muller",
    "Seltzer",
    "Shieber",
    "Smith"]


def main():

    # Prompt user for name
    name = get_string("Name: ")

    # Search for name
    left, right = 0, len(book) - 1
    while left <= right:
        middle = (left + right) // 2
        if name == book[middle]:
            print(f"Calling {name}")
            sys.exit(0)
        elif name < book[middle]:
            right = middle - 1
        elif name > book[middle]:
            left = middle + 1
    print("Quitting")
    sys.exit(1)


if __name__ == "__main__":
    main()
